### DISCUSSION TITLE
Minor warnings when building a project in platformio Arduino_GFX ver.1.5.0

### DISCUSSION THREAD

**Discussion by brightproject** [moononournation/Arduino_GFX#570] [Q&A]
My hardware and software configuration:

> PLATFORM: Espressif 32 (51.3.4+develop.sha.2656164) > Espressif ESP32-S3-DevKitC-1-N16R8V (16 MB QD, 8MB PSRAM)
> HARDWARE: ESP32S3 240MHz, 320KB RAM, 16MB Flash
> PACKAGES:
>  - framework-arduinoespressif32 @ 3.0.4
>  - framework-arduinoespressif32-libs @ 5.1.0+sha.33fbade6b8
>  - tool-esptoolpy @ 4.7.5
>  - tool-mkfatfs @ 2.0.1
>  - tool-mklittlefs @ 3.2.0
>  - tool-mkspiffs @ 2.230.0 (2.30)
>  - tool-riscv32-esp-elf-gdb @ 11.2.0+20220823
>  - tool-xtensa-esp-elf-gdb @ 11.2.0+20230208
>  - toolchain-riscv32-esp @ 12.2.0+20230208
>  - toolchain-xtensa-esp32s3 @ 12.2.0+20230208
> Dependency Graph
> |-- GFX Library for Arduino @ 1.5.0
> |-- lvgl @ 8.3.11

Declaring library settings:

> Arduino_DataBus *bus = new Arduino_ESP32QSPI(
> 10 /* CS */, 9 /* SCK */, 11 /* D0 - mosi */, 12 /* D1 - miso */, 13 /* D2 - quadwp */, 14 /* D3 - quadhd */);
> Arduino_GFX *gfx = new Arduino_ST77916(bus, 47 /* RST */, 0 /* rotation */, true /* IPS */, 360 /* width */, 360 /* height */);

I define the library initialization this way, but I get warnings when compiling the code, even if I explicitly specify the values ​​of the `SPI` bus speed.

```
void setup()
{
  /* specify data bus speed */
  gfx->begin(50000000); // 50 mHz
  // gfx->begin(16000000); // 16 mHz

  // Init Display
  // gfx->begin();
}
```

> lib/Arduino_GFX-1.5.0/src/databus/Arduino_ESP32RGBPanel.cpp: In member function 'uint16_t* Arduino_ESP32RGBPanel::getFrameBuffer(int16_t, int16_t)':
> lib/Arduino_GFX-1.5.0/src/databus/Arduino_ESP32RGBPanel.cpp:53:57: warning: narrowing conversion of '((((Arduino_ESP32RGBPanel*)this)->Arduino_ESP32RGBPanel::_prefer_speed == -1) ? ((Arduino_ESP32RGBPanel*)this)->Arduino_ESP32RGBPanel::_speed : ((Arduino_ESP32RGBPanel*)this)->Arduino_ESP32RGBPanel::_prefer_speed)' from 'int32_t' {aka 'long int'} to 'uint32_t' {aka 'long unsigned int'} [-Wnarrowing]
> Compiling .pio\build\ROUND_DISPLAY_LVGL\lib1aa\Arduino_GFX-1.5.0\databus\Arduino_ESP32SPIDMA.cpp.o
>    53 |           .pclk_hz = (_prefer_speed == GFX_NOT_DEFINED) ? _speed : _prefer_speed,
> Compiling .pio\build\ROUND_DISPLAY_LVGL\lib1aa\Arduino_GFX-1.5.0\databus\Arduino_ESP8266SPI.cpp.o
>       |                      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~Compiling .pio\build\ROUND_DISPLAY_LVGL\lib1aa\Arduino_GFX-1.5.0\databus\Arduino_HWSPI.cpp.o
> ~^~~~~~~~~~~~~~~~~~~~~~~~
> lib/Arduino_GFX-1.5.0/src/databus/Arduino_ESP32RGBPanel.cpp:54:20: warning: narrowing conversion of 'w' from 'int16_t' {aka 'short int'} to 'uint32_t' {aka 'long unsigned int'} [-Wnarrowing]
>    54 |           .h_res = w,
>       |                    ^
> lib/Arduino_GFX-1.5.0/src/databus/Arduino_ESP32RGBPanel.cpp:55:20: warning: narrowing conversion of 'h' from 'int16_t' {aka 'short int'} to 'uint32_t' {aka 'long unsigned int'} [-Wnarrowing]
>    55 |           .v_res = h,
>       |                    ^
> lib/Arduino_GFX-1.5.0/src/databus/Arduino_ESP32RGBPanel.cpp:63:56: warning: narrowing conversion of '((((int)((Arduino_ESP32RGBPanel*)this)->Arduino_ESP32RGBPanel::_hsync_polarity) == 0) ? 1 : 0)' from 'int' to 'uint32_t' {aka 'long unsigned int'} [-Wnarrowing]
>    63 |               .hsync_idle_low = (_hsync_polarity == 0) ? 1 : 0,
>       |                                 ~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
> lib/Arduino_GFX-1.5.0/src/databus/Arduino_ESP32RGBPanel.cpp:64:56: warning: narrowing conversion of '((((int)((Arduino_ESP32RGBPanel*)this)->Arduino_ESP32RGBPanel::_vsync_polarity) == 0) ? 1 : 0)' from 'int' to 'uint32_t' {aka 'long unsigned int'} [-Wnarrowing]
>    64 |               .vsync_idle_low = (_vsync_polarity == 0) ? 1 : 0,

Just an observation, when I use `Arduino_GFX` to output an image of my widget, I get `28-31 fps`

![arduino_GFX_31_fps](https://github.com/user-attachments/assets/13a255f7-90a9-4548-856a-90702595563d)

If I use libraries `ESP_PanelBus_QSPI`

> #include <ESP_IOExpander.h>
> #include <ESP_Panel_Library.h>

The same widget code gives `35-42 fps`.

![ESP_PanelBus_QSPI_44_fps](https://github.com/user-attachments/assets/976c2bd5-c49d-4bf9-b859-16511c41a052)

I'm just curious, why is there such a difference, is the `Arduino_GFX` library not optimized for the `QSPI` bus?
Please tell me @moononournation , I wanted to know your opinion as a specialist🙂

**Comment by moononournation** [comment]
yes, offical code is much faster. please stick to using that code, my library only for the beginner.

**Reply by brightproject** [reply]
> my library only for the beginner.

Thank you @moononournation  for your quick response.
Display rendering speed is important to me, but I also use your `Arduino_GFX` library quite often for learning purposes.
Could you please tell me about the minor compilation warnings described above?
I also have another question about using sprites.
Previously, I actively used the technology of obtaining an image in the form of a double buffer, including a sprite.
```
TFT_eSPI tft = TFT_eSPI();
TFT_eSprite sprite = TFT_eSprite(&tft);
```
To implement this functionality, the library described above uses only two files, which are essentially plugins

> https://github.com/Bodmer/TFT_eSPI/blob/master/Extensions/Sprite.h
> https://github.com/Bodmer/TFT_eSPI/blob/master/Extensions/Sprite.cpp

I wonder if it is possible to do this trick with the `Arduino_GFX` library?
```
Arduino_DataBus *bus = new Arduino_ESP32QSPI(
10 /* CS */, 9 /* SCK */, 11 /* D0 - mosi */, 12 /* D1 - miso */, 13 /* D2 - quadwp */, 14 /* D3 - quadhd */);
Arduino_GFX *gfx = new Arduino_ST77916(bus, 47 /* RST */, 0 /* rotation */, true /* IPS */, 360 /* width */, 360 /* height */);
TFT_eSprite sprite = TFT_eSprite(&gfx);
```
I found many similar functions for drawing in the `Arduino_GFX` and `TFT_eSPI` libraries, so I thought that there is a possibility to use the double buffer drawing functionality in the `Arduino_GFX` library.
I have several projects where I draw images on the display in sprites

```
TFT_eSPI Lcd = TFT_eSPI(); // Create an object for working with the screen

TFT_eSprite gdraw = TFT_eSprite(&Lcd); // Create a sprite for drawing
gdraw.fillCircle (px0, py0, 2 * SCREEN_Y/80, TFT_YELLOW); // 2 degree radius circle
gdraw.drawCircle (px0, py0, 2 * SCREEN_Y/80, TFT_BLACK);

gdraw.drawFastHLine (px1, py1, 3*arcSize/4, TFT_YELLOW);
gdraw.drawLine (px2, py2, px5, py5, TFT_YELLOW);
```

Please tell me, maybe the `Arduino_GFX` library already has a functionality to use a sprite similar to this design:

> gdraw.drawCircle (px0, py0, 2 * SCREEN_Y/80, TFT_BLACK);

Maybe I'll modify the code a little to add the functionality of `Sprite.cpp` to `Arduino_GFX`?

Your library works with the `Quad SPI` bus, my idea is to add the ability to work with sprites so that my projects can be displayed on displays connected via the `QSPI` bus or `RGBPanel`.
Thank you for your work🙂
